### PR TITLE
feat: kafka json append only sink

### PR DESCRIPTION
**Pull request by tabVersion**
Signed-off-by: tabVersion <tabvision@bupt.icu>

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

working on kafka sink: parse stream chunk to json and sink to kafka

we can achieve exactly once  semantic by: 
* begin a kafka transaction when a barrier comes
* kafka sink records the last success epoch num and send record requests in any epoch that is smaller or equal to the saved one will be discarded
* abort the transaction whenever a request fails then roll the system back to prev epoch
* commit to kafka if there is no failure in this epoch (the next barrier comes)

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Connector (sources & sinks)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

allow users to sink records only with `Op::Insert` to kafka in json format

the pr introduces the following options in `with` clause

* `kafka.brokers`: required, address of the Kafka broker. `Format: 'ip:port'`.
* `kafka.topic`: required, address the target topic, one sink can only have one topic 
* `sink.type`: required, accept either `append_only` or `debezium`

### CODE REVIEW

**Comment by tabVersion** [comment]
this pr has a kafka transaction issue, kafka will reject the second transaction with the error `KafkaError (Transaction error: Operation not valid in state Ready) `

File: src/connector/src/sink/mod.rs [hunk 1]
```diff
@@ -12,39 +12,83 @@
...
 use crate::sink::mysql::{MySQLConfig, MySQLSink};
 use crate::sink::redis::{RedisConfig, RedisSink};
 
 #[async_trait]
 pub trait Sink {
     async fn write_batch(&mut self, chunk: StreamChunk, schema: &Schema) -> Result<()>;
+    async fn begin_epoch(&mut self, epoch: u64) -> Result<()>;
+    async fn commit(&mut self) -> Result<()>;
+    async fn abort(&mut self) -> Result<()>;
+    async fn take_snapshot(&self) -> Result<SinkState>;
```
Comment (neverchanje) [inline src/connector/src/sink/mod.rs:40]: Needs comments above the interfaces.

File: src/connector/src/sink/kafka.rs [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+        task::spawn_blocking(move || inner.abort_transaction(timeout))
+            .unwrap_or_else(|_| Err(KafkaError::Canceled))
+    }
+
+    fn flush(&self) -> impl Future<Output = KafkaResult<()>> {
+        let inner = Arc::clone(&self.inner);
+        let timeout = self.properties.timeout;
+        task::spawn_blocking(move || inner.flush(timeout))
+            .map_ok(|_| KafkaResult::Ok(()))
+            .unwrap_or_else(|_| Err(KafkaError::Canceled))
```
Comment (neverchanje) [inline src/connector/src/sink/kafka.rs:392]: Is there any way to prevent spawn blocking?

File: src/connector/src/sink/mysql.rs [hunk 3]
```diff
@@ -40,8 +41,8 @@ pub struct MySQLSink {
 }
 
 impl MySQLSink {
-    pub fn new(cfg: MySQLConfig) -> Self {
-        Self { cfg }
+    pub fn new(cfg: MySQLConfig) -> Result<Self> {
+        Ok(Self { cfg })
```
Comment (neverchanje) [inline src/connector/src/sink/mysql.rs:45]: Why return Result when it's always Ok?

File: src/connector/src/sink/kafka.rs [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+        Ok(())
+    }
+
+    async fn abort(&mut self) -> Result<()> {
+        self.do_with_retry(|conductor| conductor.abort_transaction())
+            .await
+            .map_err(SinkError::Kafka)
+    }
+
+    async fn take_snapshot(&self) -> Result<SinkState> {
```
Comment (neverchanje) [inline src/connector/src/sink/kafka.rs:243]: What's the difference between take_snapshot and commit?

File: src/connector/src/sink/kafka.rs [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+    #[serde(rename = "kafka.topic")]
+    pub topic: String,
+
+    // Optional. If not specified, the default value is None and messages are sent to random
+    // partition. if we want to guarantee exactly once delivery, we need to specify the
+    // partition number. The partition number should set by meta.
+    pub partition: Option<i32>,
+
+    #[serde(rename = "sink.type")]
+    pub sink_type: String, // accept "append_only" or "debezium"
```
[the remark this answers is not in the document]
Comment (tabVersion) [reply src/connector/src/sink/kafka.rs:52]: I have read sink implementation on the frontend, and it looks like it is not implement it this way. We should have a doc for this.

File: src/connector/src/sink/kafka.rs (hunk 2 above)
Comment (tabVersion) [reply src/connector/src/sink/kafka.rs:392]: why? I think we can reuse more code this way

File: src/connector/src/sink/mysql.rs (hunk 3 above)
Comment (tabVersion) [reply src/connector/src/sink/mysql.rs:45]: it is for a unified interface of sink. I think we will do more checks on MySQL config later just like kafka

File: src/connector/src/sink/mod.rs (hunk 1 above)
Comment (tabVersion) [reply src/connector/src/sink/mod.rs]: fixed

File: src/stream/src/executor/sink.rs [hunk 6]
```diff
@@ -12,42 +12,76 @@
...
+                Message::Barrier(barrier) => {
+                    yield Message::Barrier(barrier);
+                }
+            }
+        }
     }
 }
 
-impl<S: Sink + 'static + Send> Executor for SinkExecutor<S> {
+impl<S: StateStore> Executor for SinkExecutor<S> {
```
Comment (nanderstabel) [inline src/stream/src/executor/sink.rs:78]: Just for my understanding, this means that SinkExecuter is a stateful operator from now on?

Comment (tabVersion) [reply src/stream/src/executor/sink.rs:78]: yes, we should store the inner state when barrier comes like other operators

**Review by nanderstabel — APPROVED** [review]

File: src/connector/src/sink/kafka.rs (hunk 4 above)
Comment (tabVersion) [reply src/connector/src/sink/kafka.rs]: remove take_snapshot func, resolve now

**Pull request merged**